src/main/java/io/jenkins/update_center/IndexHtmlBuilder.java

```diff
@@ -44,36 +53,56 @@
  */
 public class IndexHtmlBuilder implements Closeable {
     private final PrintWriter out;
+    private final String template;
+    private StringBuilder content;
+    private static String globalTemplate = null;
 
     public IndexHtmlBuilder(File dir, String title) throws IOException {
-        this(openIndexHtml(dir),title);
+        this(openIndexHtml(dir), title);
     }
 
     private static PrintWriter openIndexHtml(File dir) throws IOException {
         if (dir == null) {
             return new PrintWriter(new NullWriter()); // ignore output
         }
-        
+
         if (!dir.mkdirs() && !dir.isDirectory()) {
             throw new IllegalStateException("Failed to create " + dir);
         }
-        return new PrintWriter(new OutputStreamWriter(new FileOutputStream(new File(dir,"index.html")), StandardCharsets.UTF_8));
+        return new PrintWriter(new OutputStreamWriter(new FileOutputStream(new File(dir, "index.html")), StandardCharsets.UTF_8));
     }
 
-    public IndexHtmlBuilder(PrintWriter out, String title) {
+    private IndexHtmlBuilder(PrintWriter out, String title) {
         this.out = out;
+        if (globalTemplate == null) {
+            initTemplate();
+        }
+        template = globalTemplate.replace("{{ title }}", title)
+                .replace("{{ description }}", "Download previous versions of " + title);
+        content = new StringBuilder("<div id='grid-box'><div class=\"container\"><h1 class=\"mt-3\">")
+                .append(title).append("</h1><ul class=\"artifact-list\">\n");
+    }
+
+    private void initTemplate() {
+        Request request = new Request.Builder()
+                .url("https://www.jenkins.io/template/").get().build();
```

Comment:
so update center is supposed to be not jenkins specific right? I think it should not try and download a new template every time, but instead be something saved without the top bar.

Also is this just a single index.html? or one per directory? could it call /template a bunch of times per directory?

Reply:
One of the reasons this isn't merged yet 😉 @zbynek was just _too_ helpful after I asked for the design, I didn't get a chance to explain the implementation/configuration side of things.

Reply:
It's a lazily initiated static field, so it should be downloaded at most once per start of the Java process and reused for all the plugins. AFAIK the [bash script](https://github.com/jenkins-infra/update-center2/blob/master/site/generate.sh) starts the Jar twice, so performance should not be an issue.

>  implementation/configuration side of things

so what is the expectation here? :)

Reply:
> AFAIK the bash script starts the Jar only twice, so performance should not be an issue.

Yes, once to determine tiers, once to generate them all.

> so what is the expectation here? :)

No expectations for you, but I'm happy to accept your help 😃 My plan was sort of the following:

* New command line argument that configures the template, with the previous behavior (mostly) remaining the default.
* Add that argument to the standard invocation in the wrapper shell script so it gets applied.
* Implement template switching e.g. using an enum/strategy pattern sort of thing perhaps? Unsure in advance. Should be pretty easy to make it nicer as it is today in any case. Possibly split latest links and releases pages creation up if they don't share enough structure anymore.
* Consider moving the previous output into a "template" resource file to largely mirror the actual template file from jenkins.io except have it locally.
* Look into moving the CSS into a new HAML-based template on jenkins.io in the `templates/` folder.
* In general this is an opportunity to clean up the `IndexHtmlBuilder`. Figure out a nicer way to stitch together the output, perhaps abstract specific elements away behind a more semantic API. I basically left this class alone while I rewrote everything else earlier this year, so this is a nice opportunity to make this suck less.

(Of course some of these are really easy to defer until later, especially the cleanup oriented ones.)

Regarding template design, I think the proposed design is a huge improvement over what we currently have! A few suggestions:

* Make sure latest links (from `LatestLinkBuilder`) looks reasonable as well. Untested, but also goes through this code so may be impacted in unexpected ways.
* Use actual labels instead of (or in addition to) artifact ID. "Matrix Authorization Strategy", not (only) "matrix-auth"
* In the current design, it's a bit difficult to read the version number with the date immediately following, these should be more separated.
* Link to the plugin's plugin site page, if it's a plugin.

I played around with your design a bit, and this is what I came up with:

![Screenshot](https://user-images.githubusercontent.com/1831569/99456041-0806e680-2929-11eb-896f-046af056a883.png)

(Yes, I got my own plugin's name wrong 😄)

<details>
<summary>quick and dirty code</summary>

```
<style>.artifact-list {
  padding: 0;
}
.artifact-list .title {
  background-image: url(https://www.jenkins.io/images/jar.png);
  background-position: left center;
  background-repeat: no-repeat;
  list-style: none;
  padding-left: 2rem;
  margin-bottom: .5rem;
}
.container h2 {
    margin-top: 1rem;
}
.artifact-list td {
    padding-bottom: 0.5rem;
}
.artifact-list .title {
    font-size: 1.5rem;
  padding-right: 1rem;
}
.artifact-list .metadata {
    font-size: 0.8rem;
    color: #666;
}
</style>


<div id='grid-box'><div class="container">
    <h1 class="mt-3">Matrix Authorization Plugin</h1>
    <h3><span style="font-family: monospace;">matrix-auth</span></h3>
    <a href="https://plugins.jenkins.io/matrix-auth/">View on plugin site</a>
    <h2>Releases</h2>
    <table class="artifact-list">
        <tr>
            <td class="title" colspan="2">
                <a href='/latest/matrix-auth.hpi'>
                    Permanent to Latest Release
                </a>
            </td>
        </tr>
        <tr>
            <td class="title">
                <a href='/download/plugins/matrix-auth/2.5.1/matrix-auth.hpi'>
                    2.5.1
                </a>
            </td>
            <td class="metadata">
                    <div class="released">Released: Jun 29, 2020</div>
                    <div class="checksums">SHA-1: <code>f6046479f0a0a84de0738ac3fa144879de0f8598</code></div>
                    <div class="checksums">SHA-256: <code>7299ff13f61d4a84ccc65cea50f082bd256fd35657e16dfab081666781292c40</code></div>
            </td>
        </tr>
        <tr>
            <td class="title">
                <a href='/download/plugins/matrix-auth/2.5/matrix-auth.hpi'>
                    2.5
                </a>
            </td>
            <td class="metadata">
                    <div class="released">Released: Oct 14, 2019</div>
                    <div class="checksums">SHA-1: <code>c0d297bc809716fd3fe31427aee1391c88a59374</code></div>
                    <div class="checksums">SHA-256: <code>db99c3ecb22931ad49a81f7209c4187f2ad867f5d09ec4b4e7d19a915f543d63</code></div>
            </td>
        </tr>
    </table>
</div></div>
```

</details>

Reply:
I'd like to help with this, I'll try to follow the plan you outlined. Btw. is there a way to generate the latest links without downloading 30GB of old Jenkins versions?